src/clojure/zero_one/geni/arrow.clj

```diff
@@ -0,0 +1,222 @@
+(ns zero-one.geni.arrow
+  (:import [org.apache.arrow.vector VarCharVector
+            VectorSchemaRoot BigIntVector Float4Vector Float8Vector IntVector
+            BitVector ValueVector BaseFixedWidthVector TimeStampMilliVector]
+           [org.apache.arrow.vector.util Text]
+           [org.apache.arrow.vector.ipc ArrowStreamWriter]
+           [org.apache.arrow.memory RootAllocator]
+           [java.nio.channels Channels]
+           [org.apache.spark.sql Row]
+           [org.apache.arrow.vector.types.pojo Schema]
+           [scala.collection.convert Wrappers$IteratorWrapper]
+           )
+  (:require
+   [clojure.java.io :as io]
+
+   ))
+
+;; (set! *warn-on-reflection* true)
+
+
+;;;  these protocolls are just needed to make type hints working for the typed-action function
+(defprotocol IValueInfo
+  (^BaseFixedWidthVector vector [this])
+  (^long idx1 [this])
+  (^Object value [this])
+  )
+
+(defprotocol IRowInfo
+  (^Row row [this])
+  (^long idx2 [this])
+  )
+
+(defrecord ValueInfo [^BaseFixedWidthVector a-v ^long a-idx ^Object a-value]
+  IValueInfo
+  (vector ^BaseFixedWidthVector [this] a-v)
+  (idx1 ^long [this] a-idx)
+  (value ^Object [this] a-value)
+
+  )
+
+
+
+(defrecord RowInfo [^Row row ^long idx]
+  IRowInfo
+  (row ^Row [this] row )
+  (idx2 ^long [this] idx)
+  )
+
+;;; this method will be called for evey value transformed, so relection shoudkl be avoided
+(defn typed-action [action type value-info row-info ^String name ^RootAllocator allocator]
+  (when (= :set type)
+    (when (nil? (value value-info))
+      (throw (Exception. "value cannot be null"))
+      ))
+  (case type
+    :string (case action
+              :set (.setSafe ^VarCharVector (vector value-info) (idx1 value-info) (Text. ^String (value value-info)))
```

Comment:
I think you can make higher-level functions for `set`, `make-vector` and `get` to prevent any duplications. In particular, it's always `.set` or `setSafe` with `vector`, `idx1` and `value` called after.

Reply:
I will have a look.
I noticed the duplication, but when removing it, the type hints needed to stay. I am not super familiar with the type hints in clojure.

I was even thinking to eventually re-write the "typed-action" method in java. What do yo think ?
It is called for every single data element, so need to be fast.

Reply:
I'd rather deal with Clojure with some duplications rather than Java to be honest! I may have a crack at it!

Reply:
I discovered, that "let" allows type hints, so the protolls coul dgo away.
The code is far shorter know, and further duplication cannot be removed without giving up on type safety, which is needed in this place.